src/devices/rolling_code.c

```diff
@@ -0,0 +1,335 @@
+    uint8_t r_raw[RAW_SIZE] = {0};
+
+    // Pull out A and R bits
+    for (int i = 0; i < RAW_SIZE; i++) {
+        a_raw[i] = trinary[i * 2 + 1];
+        r_raw[i] = trinary[i * 2];
+    }       
+
+    fix_a_code(a_raw, r_raw, a_corrected);
+
+    if (debug_output > 1) {
+        char buffer[RAW_SIZE + 1];
+        raw_to_chars(a_raw, buffer, RAW_SIZE);
+        data = data_append(data, "raw_a", "", DATA_STRING, buffer, NULL);
+            raw_to_chars(r_raw, buffer, RAW_SIZE);
+        data = data_append(data, "raw_r", "", DATA_STRING, buffer, NULL);
+            raw_to_chars(a_corrected, buffer, RAW_SIZE);
+        data = data_append(data, "corrected_a", "", DATA_STRING, buffer, NULL);
+    }
+
+    if (start_width == 1) {
+        memcpy(prev_1_a_corrected, a_corrected, RAW_SIZE);
+        memcpy(prev_1_r_raw, r_raw, RAW_SIZE);
+    }
+
+    if (*prev_1_r_raw != NOT_SET && start_width == 3) {
+        char buffer[RAW_SIZE + 1];
+        uint32_t counter = get_rolling_code(
+            raw_to_uint(prev_1_r_raw, RAW_SIZE), 
+            raw_to_uint(r_raw, RAW_SIZE));
+        sprintf(buffer, "%u", counter);
+        data = data_append(data, "counter", "", DATA_STRING, buffer, NULL);
+
+        sprintf(buffer, "%08x", counter);
+        data = data_append(data, 
+            "counter_hex", "", DATA_STRING, buffer,
+            "button_pressed", "", DATA_INT, (int) a_corrected[BUTTON_TRIT],
+            "id_bits", "", DATA_INT, a_corrected[ID_TBIT_START] * 9 + a_corrected[ID_TBIT_START + 1] * 3 + a_corrected[ID_TBIT_START + 2], NULL);
+
+        uint8_t device_id[DEV_ID_SIZE];
+        memset(device_id, 0, DEV_ID_SIZE);
+        memcpy(device_id + RAW_SIZE, a_corrected, RAW_SIZE - SPECIAL_BITS);
+    
+        memcpy(device_id, prev_1_a_corrected, RAW_SIZE);
+        uint32_t value = raw_to_uint(device_id, DEV_ID_SIZE);
+        sprintf(buffer, "%08x", value);
+        data = data_append(data, 
+            "device_id", "", DATA_INT, value, 
+            "device_id_hex", "", DATA_STRING, buffer, NULL);
+    }
+
+    if (data != NULL) {
+        data = data_prepend(data, "model", "", DATA_STRING, "Rolling Code Transmitter", NULL);
+        decoder_output_data(decoder, data);
+    }
+
+    // Return 1 if message successfully decoded
+    return 1;
+}
+
+/*
+ * List of fields that may appear in the output
+ *
+ * Used to determine what fields will be output in what
+ * order for this device when using -F csv.
+ *
+ */
+static char *output_fields[] = {
+    "model",
+    "device_id",
+    "device_id_hex",
+    "counter",
+    "counter_hex",
+    "id_bits",
+    "button_pressed",
+    NULL,
+};
+
+r_device rolling_code = {
+    .name        = "Rolling Code Transmitter (-f 315M)",
+    .modulation  = OOK_PULSE_PCM_RZ,
+    .short_width = 500,  // trits are multiples of 500 uS in size
+    .long_width  = 500,  // trits are multiples of 500 uS in size
+    .reset_limit = 2000, // this is short enough so we only get 1 row
```

Comment:
3. If this limit is high enough. Both a 1 and a 3 packet might be present in the message. Then a state should not be needed.

Reply:
Please see https://github.com/merbanan/rtl_433/pull/1339/files/d21b19504eb8f2148644b6e41cb007be09ad7cb9#r396612956